File: app/src/main/java/com/example/mobdev21_night_at_the_museum/presentation/intro/LoginActivity.java
package com.example.mobdev21_night_at_the_museum.presentation.intro;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.mobdev21_night_at_the_museum.R;
import com.example.mobdev21_night_at_the_museum.presentation.home.HomeActivity;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.HashMap;
import java.util.Map;

public class LoginActivity extends AppCompatActivity {
    private static final String TAG = "Authentication";
    private static final String PREF_NAME = "MUSEUM";
    private static final String KEY_USER_EMAIL = "user_email";
    private static final String KEY_USER_PASSWORD = "user_password";
    private final String TOKEN_ID = "792900083869-59hul96s0k7e4t55k6pi15mapiicdqn9.apps.googleusercontent.com";
    private static final int RC_SIGN_IN = 9001;
    private FirebaseAuth mAuth;
    private GoogleSignInClient mGoogleSignInClient;
    private boolean show = false;
    private EditText mEmail, mPassword;
    private LinearLayout googleSignIn;
    private TextView register;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.intro2_login_fragment);
        mAuth = FirebaseAuth.getInstance();
        mEmail = findViewById(R.id.etvLoginEmail);
        mPassword = findViewById(R.id.etvLoginPassword);
        googleSignIn = findViewById(R.id.llLoginGoogle);
        register = findViewById(R.id.tvLoginSignUp);

        // Initialize Google Sign-In
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(TOKEN_ID)
                .requestEmail()
                .build();
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        // Google Sign-In button click
        googleSignIn.setOnClickListener(v -> startGoogleSignIn());

        // Email and password login button click
        findViewById(R.id.btnLogin).setOnClickListener(view -> {
            String email = mEmail.getText().toString();
            String password = mPassword.getText().toString();
            signInWithEmailAndPassword(email, password);
        });

        // Register button click
        register.setOnClickListener(v -> toRegister());
        FirebaseUser currentUser = mAuth.getCurrentUser();

        automaticSignIn();
        Log.d("SharedPreferences", "Auto Login");

    }

    private void signInWithEmailAndPassword(String email, String password) {
        mAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener(this, task -> {
            if (task.isSuccessful()) {
                saveUserLoginData(email, password);
                Log.d("SharedPreferences", "Saved User Info");
                toHomeScreen();
            } else {
                toIntroduction();
                Toast.makeText(LoginActivity.this, "Authentication failed.", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void startGoogleSignIn() {
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();
        mGoogleSignInClient.revokeAccess();
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == RC_SIGN_IN) {
            handleGoogleSignInResult(data);
        }
    }

    private void handleGoogleSignInResult(Intent data) {
        Task<GoogleSignInAccount> signInTask = GoogleSignIn.getSignedInAccountFromIntent(data);
        try {
            GoogleSignInAccount account = signInTask.getResult(ApiException.class);
            String email = account.getEmail();
            FirebaseFirestore db = FirebaseFirestore.getInstance();
            DocumentReference docRef = db.collection("users").document(email);
            docRef.get().addOnCompleteListener(task -> {
                if (task.isSuccessful()) {
                    Toast.makeText(getApplicationContext(), "User already exists", Toast.LENGTH_SHORT).show();
                } else {
                    firebaseAuthWithGoogle(account);
                }
            });
        } catch (ApiException e) {
            Log.w(TAG, "Google sign in failed", e);
        }
    }

    private void firebaseAuthWithGoogle(GoogleSignInAccount account) {
        AuthCredential credential = GoogleAuthProvider.getCredential(account.getIdToken(), null);
        mAuth.signInWithCredential(credential).addOnCompleteListener(this, authResultTask -> {
            if (authResultTask.isSuccessful()) {
                if (authResultTask.getResult().getAdditionalUserInfo().isNewUser()) {
                    createNewUserInFirestore();
                }
                toHomeScreen();
            } else {
                toIntroduction();
                Log.w(TAG, "signInWithCredential:failure", authResultTask.getException());
            }
        });
    }

    private void createNewUserInFirestore() {
        FirebaseUser user = mAuth.getCurrentUser();
        FirebaseFirestore db = FirebaseFirestore.getInstance();
        Map<String, Object> userData = new HashMap<>();
        userData.put("avatar", null);
        userData.put("fcollections", null);
        userData.put("fitems", null);
        userData.put("fstories", null);
        userData.put("name", null);
        db.collection("users").document(user.getEmail())
                .set(userData)
                .addOnSuccessListener(aVoid -> {
                    Log.d(TAG, "DocumentSnapshot successfully written!");
                    toLogin();
                })
                .addOnFailureListener(e -> {
                    Log.w(TAG, "Error writing document", e);
                    // Handle the Firestore write failure
                });
    }

    private void toRegister() {
        Intent intent = new Intent(getApplicationContext(), RegisterActivity.class);
        startActivity(intent);
    }

    private void toHomeScreen() {
        Intent homeScreen = new Intent(getApplicationContext(), HomeActivity.class);
        startActivity(homeScreen);
    }

    private void toLogin() {
        Intent toLogin = new Intent(getApplicationContext(), LoginActivity.class);
        startActivity(toLogin);
        finish();
    }

    private void toIntroduction() {
        Intent toMainScreen = new Intent(getApplicationContext(), IntroductionActivity.class);
        startActivity(toMainScreen);
        finish();
    }

    public void saveUserLoginData(String email, String password) {
        SharedPreferences sharedPreferences = getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(KEY_USER_EMAIL, email);
        editor.putString(KEY_USER_PASSWORD, password);
        editor.apply();

    }

    public void automaticSignIn() {
        SharedPreferences sharedPreferences = getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        String email = sharedPreferences.getString(KEY_USER_EMAIL, null);
        String password = sharedPreferences.getString(KEY_USER_PASSWORD, null);
        Log.d("PreferenceEmail", email);
        Log.d("PreferPass", password);
        if (email == null || password == null) {
            return;
        }
        signInWithEmailAndPassword(email, password);
    }


    public void clearUserLoginData() {
        SharedPreferences sharedPreferences = getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.remove(KEY_USER_EMAIL);
        editor.remove(KEY_USER_PASSWORD);
        editor.apply();
    }
}

File: app/src/main/java/com/example/mobdev21_night_at_the_museum/presentation/intro/LoginActivity.kt
package com.example.mobdev21_night_at_the_museum.presentation.intro

import android.text.InputType
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.ImageView
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import com.example.mobdev21_night_at_the_museum.AppPreferences
import com.example.mobdev21_night_at_the_museum.R
import com.example.mobdev21_night_at_the_museum.common.binding.MuseumActivity
import com.example.mobdev21_night_at_the_museum.common.extension.getAppFont
import com.example.mobdev21_night_at_the_museum.common.extension.getAppString
import com.example.mobdev21_night_at_the_museum.common.extension.setOnSafeClick
import com.example.mobdev21_night_at_the_museum.common.extension.toast
import com.example.mobdev21_night_at_the_museum.databinding.LoginFragment2Binding
import com.example.mobdev21_night_at_the_museum.domain.model.User
import com.example.mobdev21_night_at_the_museum.presentation.RealMainActivity
import com.example.mobdev21_night_at_the_museum.presentation.dialog.LoadingDialog2

class LoginActivity : MuseumActivity<LoginFragment2Binding>(R.layout.login_fragment_2) {

    override fun onInitView() {
        super.onInitView()
        initOnClick()
    }

    private fun initOnClick() {
        binding.tvLoginForgotPassword.setOnSafeClick { }
        binding.llLoginFacebook.setOnSafeClick { }
        binding.llLoginGoogle.setOnSafeClick { }
        binding.tvLoginSignUp.setOnSafeClick {
            navigateTo(SignUpActivity::class.java)
        }
        binding.btnLogin.setOnSafeClick { newLogin() }
        binding.etvLoginPassword.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                newLogin()
                true
            } else {
                false
            }
        }
        setPasswordEditText(binding.etvLoginPassword, binding.ivLoginShowPassword)
    }

    private fun newLogin() {
        val dialog = LoadingDialog2()
        dialog.show(supportFragmentManager, dialog::class.java.simpleName)

        val email = binding.etvLoginEmail.text.toString()
        val password = binding.etvLoginPassword.text.toString()

        val userRef = Firebase.firestore.collection("users")
        userRef.whereEqualTo("email", email).get()
            .addOnSuccessListener { querySnapshot ->
                if (querySnapshot.isEmpty || querySnapshot.documents.isEmpty()) {
                    dialog.dismiss()
                    toast(getAppString(R.string.account_does_not_exist))
                    return@addOnSuccessListener
                }
                for (document in querySnapshot.documents) {
                    val user = document.toObject(User::class.java)
                    if (user?.email == email) {
                        if (user.password == password) {
                            user.key = document.id
                            setAppPreference(user)
                            navigateToAndClearStack(RealMainActivity::class.java)
                        } else {
                            dialog.dismiss()
                            toast(getAppString(R.string.wrong_password))
                        }
                    }
                }
            }
            .addOnFailureListener { exception ->
                toast("Error getting documents: $exception")
            }
    }

    private fun login() {
        val dialog = LoadingDialog2()
        dialog.show(supportFragmentManager, dialog::class.java.simpleName)

        val email = binding.etvLoginEmail.text.toString()
        val password = binding.etvLoginPassword.text.toString()
        val ref = FirebaseDatabase.getInstance().getReference("Users")
        val query = ref.orderByChild("email").equalTo(email).limitToFirst(1)

        query.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.childrenCount == 0L) {
                    dialog.dismiss()
                    toast(getAppString(R.string.account_does_not_exist))
                    return
                }
                for (userSnapshot in snapshot.children) {
                    val user = userSnapshot.getValue(User::class.java)
                    if (user?.email == email) {
                        if (user.password == password) {
                            user.key = userSnapshot.key
                            setAppPreference(user)
                            navigateToAndClearStack(RealMainActivity::class.java)
                        } else {
                            dialog.dismiss()
                            toast(getAppString(R.string.wrong_password))
                        }
                    }
                }
            }

            override fun onCancelled(error: DatabaseError) {
                dialog.dismiss()
                toast(getAppString(R.string.sign_in_fail) + ": ${error.message}")
            }
        })
    }

    private fun setAppPreference(user: User) {
        AppPreferences.setUserInfo(user)
        AppPreferences.saveLoginInfo()
    }

    private fun setPasswordEditText(editText: EditText, imageView: ImageView) {
        imageView.setOnSafeClick {
            if (editText.inputType == InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD) {
                hidePassword(editText, imageView)
            } else {
                showPassword(editText, imageView)
            }
        }
    }

    private fun hidePassword(editText: EditText, imageView: ImageView) {
        editText.apply {
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
            setSelection(length())
            typeface = getAppFont(R.font.roboto_regular)
        }
        imageView.setImageResource(R.drawable.ic_show_password_gray)
    }

    private fun showPassword(editText: EditText, imageView: ImageView) {
        editText.apply {
            inputType = InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD
            setSelection(length())
        }
        imageView.setImageResource(R.drawable.ic_hide_password_gray)
    }
}
